projects/habitat_ovmm/evaluator.py

```diff
@@ -168,12 +137,240 @@ def stop():
+                    current_episode_metrics["goal_name"] = info["goal_name"]
+
+            metrics = self._extract_scalars_from_info(hab_info)
+            metrics_at_episode_end = {"END." + k: v for k, v in metrics.items()}
+            current_episode_metrics = {
+                **metrics_at_episode_end,
+                **current_episode_metrics,
+            }
+            current_episode_metrics["goal_name"] = info["goal_name"]
+
+            episode_metrics[current_episode_key] = current_episode_metrics
+            if len(episode_metrics) % self.metrics_save_freq == 0:
+                aggregated_metrics = self._aggregare_metrics(episode_metrics)
+                self._write_results(episode_metrics, aggregated_metrics)
+
+            count_episodes += 1
+            pbar.update(1)
+
+        self._env.close()
+        aggregated_metrics = self._aggregare_metrics(episode_metrics)
+        self._write_results(episode_metrics, aggregated_metrics)
+
+        return aggregated_metrics
+
+    def remote_evaluate(
+        self, agent, num_episodes: Optional[int] = None
+    ) -> Dict[str, float]:
+        import pickle
+        import time
+
+        import evalai_environment_habitat  # noqa: F401
+        import evaluation_pb2
+        import evaluation_pb2_grpc
+        import grpc
+
+        time.sleep(60)
+
+        def grpc_dumps(entity):
+            return pickle.dumps(entity)
+
+        def grpc_loads(entity):
+            return pickle.loads(entity)
+
+        env_address_port = os.environ.get("EVALENV_ADDPORT", "localhost:8085")
+        channel = grpc.insecure_channel(env_address_port)
+        stub = evaluation_pb2_grpc.EnvironmentStub(channel)
+
+        env_num_episodes = grpc_loads(
+            stub.number_of_episodes(evaluation_pb2.Package()).SerializedEntity
+        )
+        if num_episodes is None:
+            num_episodes = env_num_episodes
+        else:
+            assert num_episodes <= env_num_episodes, (
+                "num_episodes({}) is larger than number of episodes "
+                "in environment ({})".format(num_episodes, env_num_episodes)
+            )
+
+        assert num_episodes > 0, "num_episodes should be greater than 0"
+
+        episode_metrics: Dict = {}
+
+        count_episodes: int = 0
+
+        pbar = tqdm(total=num_episodes)
+        while count_episodes < num_episodes:
+            observations, done = (
+                grpc_loads(stub.reset(evaluation_pb2.Package()).SerializedEntity),
+                False,
+            )
+            current_episode = grpc_loads(
+                stub.get_current_episode(evaluation_pb2.Package()).SerializedEntity
+            )
+            agent.reset_vectorized([current_episode])
+
+            current_episode_key = (
+                f"{current_episode.scene_id.split('/')[-1].split('.')[0]}_"
+                f"{current_episode.episode_id}"
+            )
+            current_episode_metrics = {}
+
+            while not done:
```

Comment:
Per episode - might be cleaner if separate function?

Reply:
This is commonly used habitat style loop, so I'd better leave it as it is.